src/sentry/incidents/models/alert_rule.py

```diff
@@ -550,6 +550,20 @@ def build_handler(
         return self.trigger_action_class(action, incident, project)
 
 
+class FactoryRegistry:
+    def __init__(self) -> None:
+        self.by_action_service: dict[ActionService, ActionHandlerFactory] = {}
+        self.by_slug: dict[str, ActionHandlerFactory] = {}
+
+    def register(self, factory: ActionHandlerFactory) -> None:
+        if factory.service_type in self.by_action_service:
+            raise Exception(f"Handler already registered for type {factory.service_type}")
+        if factory.slug in self.by_slug:
+            raise Exception(f"Handler already registered with slug={factory.slug!r}")
```

Comment:
what does the `!r` do here?

Reply:
In an f-string, it's shorthand for `f"slug={repr(factory.slug)}"`, so that it prints `slug="slack"` rather than `slug=slack`.